apollo-router/src/axum_factory/connection_handle.rs

```diff
@@ -0,0 +1,88 @@
+use std::collections::HashMap;
+use std::sync::OnceLock;
+
+use parking_lot::Mutex;
+use parking_lot::MutexGuard;
+
+use crate::ListenAddr;
+use crate::services::router::pipeline_handle::PipelineRef;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) enum ConnectionState {
+    Active,
+    Terminating,
+}
+
+/// A ConnectionRef is used to keep track of how many connections we have active. It's associated with an instance of RouterCreator
+/// Pipeline ref represents a unique pipeline
+#[derive(Clone, Hash, Eq, PartialEq, Debug)]
+pub(crate) struct ConnectionRef {
+    pub(crate) pipeline_ref: PipelineRef,
+    pub(crate) address: ListenAddr,
+    /// The state of this connection. When we are trying to shut it down, for instance on reload, it will switch to terminating.
+    pub(crate) state: ConnectionState,
+}
+
+/// A connection handle does the actual tracking of connections
+/// Creating a new connection handle will insert a ConnectionRef into a static map.
+/// Dropping all connection handles associated with the internal ref will remove the ConnectionRef
+/// Clone MUST NOT be implemented for this type. Cloning will make extra copies that when dropped will throw off the global count.
+pub(crate) struct ConnectionHandle {
+    pub(crate) connection_ref: ConnectionRef,
+}
+
+static CONNECTIONS: OnceLock<Mutex<HashMap<ConnectionRef, u64>>> = OnceLock::new();
+pub(crate) fn connections() -> MutexGuard<'static, HashMap<ConnectionRef, u64>> {
```

Comment:
Reading the code elsewhere in the PR I thought this would iterate over *all* open connections (meaning potentially many). Maybe this could be named `connection_counts()` or at least a doc comment added to explain what it really is?

Reply:
[c9e0e31](https://github.com/apollographql/router/pull/7009/commits/c9e0e31f44782fc95c3ec51abc7d1f8ea9726735)